docs/madr/decisions/006-mesh-traffic-permission.md

```diff
@@ -0,0 +1,785 @@
+
+* Status: accepted
+
+Technical Story: https://github.com/kumahq/kuma/issues/4222
+
+## Context and Problem Statement
+
+[New policy matching MADR](./005-policy-matching.md) introduces a new approach how Kuma applies configuration to proxies.
+Rolling out strategy implies creating a new set of policies that satisfy a new policy matching.
+Current MADR aims to define a MeshTrafficPermission.
+
+## Considered Options
+
+* Create a MeshTrafficPermission policy
+
+## Decision Outcome
+
+Chosen option: create a MeshTrafficPermission policy
+
+### Overview
+
+MeshTrafficPermission allows blocking unwanted incoming traffic for the service.
+There are might be several reasons why it could be useful to block unwanted traffic:
+* following the [principle of least privilege](https://en.wikipedia.org/wiki/Principle_of_least_privilege)
+* cut off access for clients that shouldn't use your service (legacy clients, clients from another zone, etc.)
+
+### Specification
+
+Top-level targetRef can have all available kinds:
+
+```yaml
+targetRef:
+  kind: Mesh|MeshSubset|MeshService|MeshServiceSubset|MeshGatewayRoute|MeshHTTPRoute
+  name: ...
+```
+
+MeshTrafficPermission is an inbound policy that's why it has a "from" array:
+
+```yaml
+from:
+  - targetRef:
+      kind: Mesh|MeshSubset|MeshService|MeshServiceSubset
+      name: ...
+```
+
+Configuration for MeshTrafficPermission has a single field "action" with values "ALLOW", "DENY", 
+"ALLOW_WITH_SHADOW_DENY" and "DENY_WITH_SHADOW_ALLOW".
+So the overall specification looks like this:
+
+```yaml
+type: MeshTrafficPermission
+mesh: default
+spec:
+  targetRef:
+    kind: Mesh|MeshSubset|MeshService|MeshServiceSubset|MeshGatewayRoute|MeshHTTPRoute
+    name: ...
+  from:
+    - targetRef:
+        kind: Mesh|MeshSubset|MeshService|MeshServiceSubset
+        name: ...
+      default:
+        action: ALLOW|DENY|ALLOW_WITH_SHADOW_DENY|DENY_WITH_SHADOW_ALLOW
+```
+
+We don't want to mix mTLS configs (like permissive mTLS or TLS version) with MeshTrafficPermission. 
+The main reason is that mTLS settings are applied for the entire inbound, while permissions could be set for a group of clients. 
+
+### Shadow actions
+
+Shadow actions are not enforced but emit stats and can be used for rule testing.
+For example, before creating "action: DENY" rule, user creates "action: ALLOW_WITH_SHADOW_DENY" rule and checks the stats.
+If there are no unexpected denied requests from the clients then it's safe to change "action" to "DENY".
+
+#### Other considered options
+
+Option 1: `action: ALLOW|DENY, shadow: true|false`
+
+Cons: policy matching makes it less obvious with merging. It's easy to create a rule like:
+```yaml
+targetRef:
+   kind: Mesh
+default:
+  action: DENY
+  shadow: true
+```
+and unintentionally shadow all less specific rules.
+
+Option 2: `action: ALLOW|DENY, shadowAction: ALLOW|DENY`
+
+Cons:
+* there are unused states like `action: ALLOW, shadowAction: ALLOW` and `action: DENY, shadowAction: DENY`
```

Comment:
I don't think `action: DENY, shadowAction: DENY` is a useless state, dynamic metadata and access log hints are only available for the shadow engine AFAIK. See my previous comment.

Reply:
> dynamic metadata and access log hints are only available

can you expand on what these are and why they might be needed?

Reply:
dynamic metadata is data emitted by filters via `setDynamicMetadata` that can be used down the chain in other filters (see [this](https://www.envoyproxy.io/docs/envoy/latest/configuration/advanced/well_known_dynamic_metadata)). 

access log hints - This is how an RBAC filter can signal (via dynamic metadata) to other filter that this request should be logged. Side note: this is something that was **not** implemented in Envoy when I worked on TrafficPermissions in Allegro (so I might be wrong in some places).

When we sat down to design TrafficPermissions in Allegro we thought about the end goal of this feature but also about the migration. How would a pre-existing application go from no TrafficPermissions to being "locked-down" via Traffic permissions safely? First of all you need a "grace period" in which the traffic is allowed and metrics / logs are emitted that help to build TrafficPermissions rules. In allegro this was done via logs (that's why I mention access log hints) and stats (we don't have equivalent now I think).

I'm not saying that this feature cannot be released without this but I would be careful about going through any "one way doors" that could significantly complicate implementing things mentioned above.

Reply:
Do remember we can just add `DENY_AND_SHADOW_DENY`/`ALLOW_AND_SHADOW_ALLOW` at any time.

Reply:
Is this the same with ALLOW/DENY/**LOG**?

Reply:
Sure, `DENY_AND_SHADOW_LOG` :shrug: 

Btw I think the `access_log_hint` is supported for both actions and shadow actions.